config/boot.rb

```diff
@@ -1,3 +1,8 @@
+# Set up gems listed in the Gemfile.
 ENV['BUNDLE_GEMFILE'] ||= File.expand_path('../../Gemfile', __FILE__)
 
 require 'bundler/setup' # Set up gems listed in the Gemfile.
+
+# load dotenv files before "before_configuration" callback
+require File.expand_path('../config_root', __FILE__)
+ConfigRoot.load_dotenv_files
```

Comment:
Maybe I forgot, but what is the reason for including these two lines in `boot.rb` as opposed to the head of the `configuration.rb` file?

Reply:
So you'd have something like...

```ruby
# config/configuration.rb
require File.expand_path('../config_root', __FILE__)
ConfigRoot.load_dotenv_files

# dashboard app specific configuration
class Configuration
  extend ConfigRoot

  ...
end

...
```

Reply:
It could be shortened further by calling `ConfigRoot.load_dotenv_files` at the end of `config/config_root.rb` instead of outside of it and after the `require`.

Reply:
So that would look like:

```ruby
# config/config_root.rb
require 'pathname'
require 'dotenv'

module ConfigRoot
  ...
end

ConfigRoot.load_dotenv_files
```

and

```ruby
# config/configuration.rb
require File.expand_path('../config_root', __FILE__)

# dashboard app specific configuration
class Configuration
  extend ConfigRoot

  ...
end

# support custom initializers and views in /etc
Rails.application.configure do |config|
  config.paths["config/initializers"] << Configuration.config_root.join("config", "initializers").to_s
  config.paths["app/views"].unshift Configuration.config_root.join("app", "views").to_s
end
```

If this were Job Composer you would then have a `bin/setup-production` with...

```ruby
#!/usr/bin/env ruby
ENV['RAILS_ENV'] = 'production'  # run this in production

require 'pathname'
require 'fileutils'
require_relative '../config/config_root.rb'

...
```

Reply:
It is required in boot because we have to

1. require it before before_configuration callback is called
2. require before all of the Rails application code is initialized to increase parity between the loading this file when launching an app, versus running a rake task, versus requiring in `bin/setup-production` (https://github.com/OSC/ood-myjobs/blob/eaf1085344886b6655bf55fd29b8a095c6348ac8/bin/setup-production#L9 is what bin/rake does so I do that instead of `require_relative '../config/config_root'
` which I was doing before)

I'm warming to the idea of merging config_root and configuration, instead of keeping separate. We'd move the rails config.paths initialization back to the application.rb then. But we would still load it in boot.rb, not in environment.rb.

Reply:
Ah, makes sense putting it in `boot.rb` then.

> I'm warming to the idea of merging config_root and configuration, instead of keeping separate. We'd move the rails config.paths initialization back to the application.rb then. But we would still load it in boot.rb, not in environment.rb.

I am for this.